Perl/Decoder/t/091_thaw_order.t

```diff
@@ -0,0 +1,129 @@
+            inside => {
+                map { $_ => $inside[$_] } 0..$#inside
+            },
+        }, $class;
+    }
+
+
+    sub FREEZE {
+        my ( $self, $serializer ) = @_;
+
+        push @steps, 'FooOutside::FREEZE(' . ::dd($self) . ')';
+        return {
+            attrs  => $self->{attrs},
+            inside => [
+                map { $self->{inside}{$_} } sort { $a <=> $b } keys %{$self->{inside}},
+            ],
+        };
+    }
+
+    sub THAW {
+        my ( $class, $serializer, @data ) = @_;
+
+        push @steps, 'FooOutside::THAW(' . ::dd($data[0]) . ')';
+        return $class->new( $data[0]->{attrs}, @{ $data[0]->{inside} || [] } );
+    }
+}
+
+my $struct = 
+    FooOutside->new(
+        # Attrs
+        {
+            attr1 => 'foobar',
+            bar => FooInside->new({}, 'should_be_first'),
+        },
+        FooInside->new({}, 'second/not_frozen', 'third/not_frozen'),
+        FooOutside->new({}, FooInside->new({}, 'fourth/inner')),
+    );
+
+my $srl = Sereal::Encoder::encode_sereal($struct, { freeze_callbacks => 1, canonical => 1 });
+my $dec = decode_sereal($srl);
+
+
+is_deeply(\@steps, [
+          "FooOutside::FREEZE(bless( {'attrs' => {'attr1' => 'foobar','bar' => bless( {'attrs' => {},'inside' => {'0' => 'should_be_first'}}, 'FooInside' )},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'second/not_frozen','1' => 'third/not_frozen'}}, 'FooInside' ),'1' => bless( {'attrs' => {},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )}}, 'FooOutside' )}}, 'FooOutside' ))",
+          "FooInside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => 'should_be_first'}}, 'FooInside' ))",
+          "FooInside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => 'second/not_frozen','1' => 'third/not_frozen'}}, 'FooInside' ))",
+          "FooOutside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )}}, 'FooOutside' ))",
+          "FooInside::FREEZE(bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' ))",
+          "FooInside::THAW({'attrs' => {},'inside' => ['fourth/inner']})",
+          "FooOutside::THAW({'attrs' => {},'inside' => [bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )]})",
+          "FooInside::THAW({'attrs' => {},'inside' => ['second/not_frozen','third/not_frozen']})",
+          "FooInside::THAW({'attrs' => {},'inside' => ['should_be_first']})",
+          "FooOutside::THAW({'attrs' => {'attr1' => 'foobar','bar' => bless( {'attrs' => {},'inside' => {'0' => 'should_be_first'}}, 'FooInside' )},'inside' => [bless( {'attrs' => {},'inside' => {'0' => 'second/not_frozen','1' => 'third/not_frozen'}}, 'FooInside' ),bless( {'attrs' => {},'inside' => {'0' => bless( {'attrs' => {},'inside' => {'0' => 'fourth/inner'}}, 'FooInside' )}}, 'FooOutside' )]})"
+        ], "freeze/thaw sequence for nested objects was as expected") or do {
```

Comment:
The sequence looks a bit odd as it's tracking what is getting frozen (i.e. the object) vs what is getting thawed (i.e. as of yet a likely intermediate representation).

Assuming there exists something like `is_deeply` that can properly deal with nested objects and can check for "the right" blessedness of them however deep, would it be possible to have a test that verifies that the result of the decode is indeed "in the same shape" as the encoded one?

The "are the steps like I think they should be" is not a good substitute for making sure that the two objects (before and after) have the same shape.

Reply:
> But regardless, adding a specific test that the final decoded object is the same as the original is easy, and done now.

Beware this:

```bash
$ perl -MTest::More -E'say is_deeply({foo=>1},bless{foo=>1},"Bar");done_testing;'
ok 1
1
1..1

$ perl -MData::Dumper::Concise=Dumper -MTest::More -E'say is_deeply(Dumper({foo=>1}),Dumper(bless{foo=>1},"Bar"));done_testing;'
not ok 1
#   Failed test at -e line 1.
#          got: '{
#   foo => 1,
# }
# '
#     expected: 'bless( {
#   foo => 1,
# }, 'Bar' )
# '
0
1..1
# Looks like you failed 1 test of 1.```

I'll soon be able to have a look and more properly check that this fixes my problem well. Thanks for your help so far!

> you did a patch to solve this, and wrote a test which passed, but then you said that your original script showed something wrong

Yes, it purportedly "fixed" the order in a naive case, but didn't fix it deeply: the ordering of the array wasn't right, so swapping it didn't fix much.

> So i wanted to double check that in practice the fixed code was working as you expected. I am relatively confident this fix is the correct fix, but i wanted to verify that once you use this patch it works as you expect.

Sounds good. Thanks for that. I will be able to test this soon and I'll report back.

Thanks for your help so far!